src/core/socket.c

```diff
@@ -2542,6 +2613,77 @@ _pure_ static bool socket_check_gc(Unit *u) {
         return s->n_connections > 0;
 }
 
+SocketPeer *socket_peer_new(void) {
+        SocketPeer *p;
+
+        p = new0(SocketPeer, 1);
+        if (!p)
+                return NULL;
+
+        return p;
+}
+
+SocketPeer *socket_peer_ref(SocketPeer *p) {
+        if (!p)
+                return NULL;
+
+        p->n_ref++;
+
+        return p;
+}
+
+SocketPeer *socket_peer_unref(SocketPeer *p) {
+        if (!p)
+                return NULL;
+
+        assert(p->n_ref > 0);
+
+        p->n_ref--;
+
+        if (p->n_ref == 0)
+                free(p);
+
+        return NULL;
+}
+
+int socket_find_peer(Socket *s, int fd, SocketPeer **p) {
+        _cleanup_free_ SocketPeer *remote = NULL;
+        SocketPeer *i;
+        socklen_t l;
+        int r;
+
+        assert(fd >= 0);
+        assert(s);
+
+        remote = socket_peer_new();
+        if (!remote)
+                return log_oom();
+
+        l = sizeof(remote->peer.sa);
+        r = getpeername(fd, (struct sockaddr*) &remote->peer.sa, &l);
+        if (r < 0) {
+                log_error_errno(errno, "getpeername failed: %m");
+                return -errno;
+        }
+
+        remote->family = remote->peer.sa.sa_family;
+
+        i = hashmap_get(s->peers_by_address, remote);
+        if (i) {
+                *p = i;
+                return 1;
+        }
+
+        r = hashmap_put(s->peers_by_address, remote, remote);
+        if (r < 0)
+                return r;
```

Comment:
Oh. You are right. We need the family hence sockaddr_union is the right choice. Sorry for the confusion!